hexkit/protocols/objstorage.py

```diff
@@ -0,0 +1,639 @@
+    # (is typically only used by the protocol but may also be used in
+    # provider-specific code)
+
+    @classmethod
+    def _validate_bucket_id(cls, bucket_id: str):
+        """Check whether a bucket id follows the recommended naming pattern.
+        This is roughly based on:
+        https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html
+        Raises BucketIdValidationError if not valid.
+        """
+        if len(bucket_id) not in range(3, 64):
+            raise cls.BucketIdValidationError(
+                bucket_id=bucket_id,
+                reason="must be between 3 and 63 character long",
+            )
+        if not re.match(r"^[a-z0-9\-]*$", bucket_id):
+            raise cls.BucketIdValidationError(
+                bucket_id=bucket_id,
+                reason="only lowercase letters, numbers, and hyphens (-) are allowd",
+            )
+        if bucket_id.startswith("-") or bucket_id.endswith("-"):
+            raise cls.BucketIdValidationError(
+                bucket_id=bucket_id,
+                reason="may not start or end with a hyphen (-).",
+            )
+
+    @classmethod
+    def _validate_object_id(cls, object_id: str):
+        """Check whether a object id follows the recommended naming pattern.
+        This is roughly based on (plus some additional restrictions):
+        https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-keys.html
+        Raises ObjectIdValidationError if not valid.
+        """
+        if len(object_id) not in range(3, 64):
+            raise cls.ObjectIdValidationError(
+                object_id=object_id,
+                reason="must be between 3 and 63 character long",
+            )
+        if not re.match(r"^[a-zA-Z0-9\-\.]*$", object_id):
+            raise cls.ObjectIdValidationError(
+                object_id=object_id,
+                reason="only letters, numbers, and hyphens (-), and dots (.) are allowd",
+            )
+        if re.match(r"^[\-\.].*", object_id) or re.match(r".*[\-\.]$", object_id):
+            raise cls.ObjectIdValidationError(
+                object_id=object_id,
+                reason="may not start or end with a hyphen (-) or a dot (.).",
+            )
+
+    # Exceptions that may be used by implementation:
+
+    class ObjectStorageProtocolError(RuntimeError):
+        """Generic base exceptions for all custom errors used by this protocol."""
+
+    class BucketError(ObjectStorageProtocolError):
+        """Generic base exceptions for error that occur while handling buckets."""
+
+    class BucketNotFoundError(BucketError):
+        """Thrown when trying to access a bucket with an ID that doesn't exist."""
+
+        def __init__(self, bucket_id: Optional[str]):
+            message = (
+                "The bucket "
+                + (f"with ID '{bucket_id}' " if bucket_id else "")
+                + "does not exist."
+            )
+            super().__init__(message)
```

Comment:
```suggestion
            with_id = f" with ID '{bucket_id}'" if bucket_id else ""
            )
            super().__init__(f"The bucket{with_id} does not exist.")
```
Similarly below, if you like that style.